aries_cloudagent/messaging/jsonld/create_verify_data.py

```diff
@@ -35,59 +37,78 @@ def _cannonize_document(doc):
-            datetime.timezone.utc
-        ).strftime("%Y-%m-%dT%H:%M:%SZ")
+    def created_at():
+        stamp = datetime.datetime.now(datetime.timezone.utc)
+        return stamp.strftime("%Y-%m-%dT%H:%M:%SZ")
 
-    if (
-        "type" not in signature_options
-        or signature_options["type"] != "Ed25519Signature2018"
-    ):
-        signature_options["type"] = "Ed25519Signature2018"
+    signature_options["created"] = signature_options.get("created", created_at())
 
-    [expanded] = jsonld.expand(data)
+    # TODO: povide methods for jsonld expand that reports dropped attributes for better error reporting
+    [expanded] = jsonld.expand(data, {"keepFreeFloatingNodes": True})
     framed = jsonld.compact(
         expanded, "https://w3id.org/security/v2", {"skipExpansion": True}
     )
-
     # Detect any dropped attributes during the expand/contract step.
-
-    if len(data) != len(framed):
-        raise DroppedAttributeException("Extra Attribute Detected")
-    if (
-        "proof" in data
-        and "proof" in framed
-        and len(data["proof"]) != len(framed["proof"])
-    ):
-        raise DroppedAttributeException("Extra Attribute Detected")
-    if (
-        "credentialSubject" in data
-        and "https://www.w3.org/2018/credentials#credentialSubject" in framed
-        and len(data["credentialSubject"])
-        != len(framed["https://www.w3.org/2018/credentials#credentialSubject"])
-    ):
-        raise DroppedAttributeException("Extra Attribute Detected")
+    if len(data) > len(framed):
+        # attempt to collect error report data
+        for_diff = jsonld.compact(expanded, data.get("@context"))
+        dropped = set(data.keys()) - set(for_diff.keys())
+        raise DroppedAttributeException(
+            f"{dropped} attributes dropped. "
+            " Provide definitions in context to correct."
+        )
+    # Check proof for dropped attributes
+    attr = "proof"
+    if len(data.get(attr, {})) > len(framed.get(attr, {})):
+        for_diff = jsonld.compact(expanded, data.get("@context"))
+        dropped = set(data.get(attr, {}).keys()) - set(for_diff.get(attr, {}).keys())
+        raise DroppedAttributeException(
+            f"in {attr}, {dropped} attributes dropped. "
+            "Provide definitions in context to correct."
+        )
+    # Check credentialSubject for dropped attributes
+    attr = [
+        "credentialSubject",
+        "https://www.w3.org/2018/credentials#credentialSubject",
+    ]
+    if len(data.get(attr[0], {})) > len(framed.get(attr[1], {})):
+        for_diff = jsonld.compact(expanded, data.get("@context"))
+        dropped = set(data.get(attr[0], {}).keys()) - set(
+            for_diff.get(attr[1], {}).keys()
+        )
+        raise DroppedAttributeException(
+            f"in {attr[0]}, {dropped} attributes dropped."
+            "Provide definitions in context to correct."
+        )
+        raise DroppedAttributeException("CredentialSubject dropped")
```

Comment:
Should be able to simplify this by iterating through a list that looks like:
```python
attrs = [("proof", "proof"), ("credentialSubject", "https://www.w3.org/....#credentialSubject")]
```

Setting intermediate values for the `get`s would also help with readability, I think. There's a lot of characters in there so having some of that unraveled a bit would make it easier to parse.

Reply:
good idea, I will make the change. I also will remove the trailing exception on line 111.